File: android/lib/model/finance_transaction.dart
import 'package:android/model/id_aware.dart';
import 'package:json_annotation/json_annotation.dart';

class FinanceTransaction extends IdAware<String> {
  String id;
  String sourceDepositId;
  String targetDepositId;
  FinanceTransactionState state;
  double value;
  DateTime executionDatetime;
  String creatorId;
  String name;
  String note;
  Set<String> categoriesIds;

  FinanceTransaction(this.id);

  @override
  String toString() {
    return 'FinanceTransaction{id: $id, sourceDepositId: $sourceDepositId, '
        'targetDepositId: $targetDepositId, state: $state, value: $value,'
        ' executionDatetime: $executionDatetime, creatorId: $creatorId, name: '
        '$name, note: $note, categoriesIds: $categoriesIds}';
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is FinanceTransaction &&
          runtimeType == other.runtimeType &&
          id == other.id;

  @override
  int get hashCode => id.hashCode;
}

enum FinanceTransactionState {
  @JsonValue('PLANNED')
  planned,
  @JsonValue('BLOCKED')
  blocked,
  @JsonValue('CANCELLED')
  cancelled,
  @JsonValue('EXECUTED')
  executed
}

File: server/src/main/kotlin/eu/mlich/lucy/model/finance/FinanceTransaction.kt
package eu.mlich.lucy.model.finance

import eu.mlich.lucy.model.User
import java.time.LocalDateTime
import java.util.*
import javax.persistence.*

/**
 * Represent a transaction between a [FinanceDeposit] or anything external.
 *
 * @author Juraj Mlich <jurajmlich@gmail.com>
 */
@Entity
@Table(name = "finance_transaction")
data class FinanceTransaction(
        @Id
        @GeneratedValue(strategy = GenerationType.IDENTITY)
        @Column(name = "id")
        var id: Int? = null,

        /**
         * Originating deposit. Can be null in case it is external entity that is not managed by the application.
         */
        @ManyToOne
        @JoinColumn(name = "source_deposit_id")
        var sourceDeposit: FinanceDeposit?,

        @ManyToOne
        @JoinColumn(name = "target_deposit_id")
        var targetDeposit: FinanceDeposit?,

        @Column(name = "state")
        @Enumerated(EnumType.STRING)
        var state: FinanceTransactionState,

        @Column(name = "value")
        var value: Double,

        @Column(name = "execution_datetime")
        var executionDatetime: LocalDateTime?,

        @ManyToOne
        @JoinColumn(name = "creator_id")
        val creator: User,

        @Column(name = "name")
        var name: String? = null,

        @Column(name = "note")
        var note: String? = null,

        @ManyToMany
        @JoinTable(
                name = "finance_transaction_transaction_category",
                joinColumns = [JoinColumn(name = "transaction_id")],
                inverseJoinColumns = [(JoinColumn(name = "transaction_category_id"))]
        )
        var categories: Set<FinanceTransactionCategory> = HashSet(),

        @Column(name = "public_key")
        var publicKey: UUID = UUID.randomUUID()

) {
        override fun equals(other: Any?) = other is FinanceTransaction && other.publicKey === publicKey

        override fun hashCode(): Int = publicKey.hashCode()
}
